src/lib/message/ExchangeContext.cpp

```diff
@@ -0,0 +1,1664 @@
+    // the application may have released its reference, resulting in the exchange context actually
+    // being freed here.
+    Release();
+
+    return err;
+}
+
+/**
+ *  Send a Common::Null message.
+ *
+ *  @note  When sent via UDP, the null message is sent *without* requesting an acknowledgment,
+ *  even in the case where the auto-request acknowledgment feature has been enabled on the
+ *  exchange.
+ *
+ *  @retval  #CHIP_ERROR_NO_MEMORY   If no available PacketBuffers.
+ *  @retval  #CHIP_NO_ERROR          If the method succeeded or the error wasn't critical.
+ *  @retval  other                    Another critical error returned by SendMessage().
+ *
+ */
+CHIP_ERROR ExchangeContext::SendCommonNullMessage(void)
+{
+    CHIP_ERROR err        = CHIP_NO_ERROR;
+    PacketBuffer * msgBuf = NULL;
+
+    // Allocate a buffer for the null message
+    msgBuf = PacketBuffer::NewWithAvailableSize(0);
+    VerifyOrExit(msgBuf != NULL, err = CHIP_ERROR_NO_MEMORY);
+
+    // Send the null message
+    err =
+        SendMessage(chip::Profiles::kChipProfile_Common, chip::Profiles::Common::kMsgType_Null, msgBuf, kSendFlag_NoAutoRequestAck);
+    msgBuf = NULL;
+
+exit:
+    if (ChipMessageLayer::IsSendErrorNonCritical(err))
+    {
+        ChipLogError(ExchangeManager, "Non-crit err %ld sending solitary ack", long(err));
+        err = CHIP_NO_ERROR;
+    }
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(ExchangeManager, "Failed to send Solitary ack for MsgId:%08" PRIX32 " to Peer %016" PRIX64 ":%ld",
+                     mPendingPeerAckId, PeerNodeId, (long) err);
+    }
+
+    return err;
+}
+
+/**
+ *  Encode the exchange header into a message buffer.
+ *
+ *  @param[in]    exchangeHeader     A pointer to the CHIP Exchange header object.
+ *
+ *  @param[in]    profileId          The profile identifier of the CHIP message to be sent.
+ *
+ *  @param[in]    msgType            The message type of the corresponding profile.
+ *
+ *  @param[in]    msgBuf             A pointer to the PacketBuffer needed for the CHIP message.
+ *
+ *  @param[in]    sendFlags          Flags set by the application for the CHIP message being sent.
+ *
+ *
+ *  @retval  #CHIP_ERROR_BUFFER_TOO_SMALL  If the message buffer does not have sufficient space
+ *                                          for encoding the exchange header.
+ *  @retval  #CHIP_NO_ERROR                If encoding of the message was successful.
+ */
+CHIP_ERROR ExchangeContext::EncodeExchHeader(ChipExchangeHeader * exchangeHeader, uint32_t profileId, uint8_t msgType,
+                                             PacketBuffer * msgBuf, uint16_t sendFlags)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // Fill the exchange header to the message buffer.
+    exchangeHeader->Version     = kChipExchangeVersion_V1;
+    exchangeHeader->ExchangeId  = ExchangeId;
+    exchangeHeader->ProfileId   = profileId;
+    exchangeHeader->MessageType = msgType;
```

Comment:
How this interacts with existing CHIP security and protocolID, isn't clear to me... might be worth having folks sync up to understand how this plans to land in the existing CHIP codebase?